storage/remote/write_handler.go

```diff
@@ -135,114 +102,66 @@ func isHistogramValidationError(err error) bool {
-	if enc == "" {
-		// Don't break yolo 1.0 clients if not needed. This is similar to what we did
-		// before 2.0: https://github.com/prometheus/prometheus/blob/d78253319daa62c8f28ed47e40bafcad2dd8b586/storage/remote/write_handler.go#L62
-		// We could give http.StatusUnsupportedMediaType, but let's assume snappy by default.
-	} else if strings.ToLower(enc) != compression.Snappy {
-		err := fmt.Errorf("%v encoding (compression) is not accepted by this server; only %v is acceptable", enc, compression.Snappy)
-		h.logger.Error("Error decoding remote write request", "err", err)
-		http.Error(w, err.Error(), http.StatusUnsupportedMediaType)
-		return
-	}
-
-	// Read the request body.
+// Store implements remoteapi.writeStorage interface.
+func (h *writeHandler) Store(r *http.Request, msgType remoteapi.WriteMessageType) (*remoteapi.WriteResponse, error) {
+	// Store receives request with snappy-decompressed content in body.
 	body, err := io.ReadAll(r.Body)
 	if err != nil {
-		h.logger.Error("Error decoding remote write request", "err", err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
-	}
-
-	decompressed, err := compression.Decode(compression.Snappy, body, nil)
-	if err != nil {
-		// TODO(bwplotka): Add more context to responded error?
-		h.logger.Error("Error decompressing remote write request", "err", err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		h.logger.Error("Error reading remote write request body", "err", err.Error())
+		return nil, err
 	}
 
-	// Now we have a decompressed buffer we can unmarshal it.
-
-	if msgType == config.RemoteWriteProtoMsgV1 {
+	wr := remoteapi.NewWriteResponse()
+	if msgType == remoteapi.WriteV1MessageType {
 		// PRW 1.0 flow has different proto message and no partial write handling.
 		var req prompb.WriteRequest
-		if err := proto.Unmarshal(decompressed, &req); err != nil {
+		if err := proto.Unmarshal(body, &req); err != nil {
 			// TODO(bwplotka): Add more context to responded error?
 			h.logger.Error("Error decoding v1 remote write request", "protobuf_message", msgType, "err", err.Error())
-			http.Error(w, err.Error(), http.StatusBadRequest)
-			return
+			wr.SetStatusCode(http.StatusBadRequest)
+			return wr, err
 		}
 		if err = h.write(r.Context(), &req); err != nil {
 			switch {
 			case errors.Is(err, storage.ErrOutOfOrderSample), errors.Is(err, storage.ErrOutOfBounds), errors.Is(err, storage.ErrDuplicateSampleForTimestamp), errors.Is(err, storage.ErrTooOldSample):
 				// Indicated an out-of-order sample is a bad request to prevent retries.
-				http.Error(w, err.Error(), http.StatusBadRequest)
-				return
+				wr.SetStatusCode(http.StatusBadRequest)
+				return wr, err
 			case isHistogramValidationError(err):
-				http.Error(w, err.Error(), http.StatusBadRequest)
-				return
+				wr.SetStatusCode(http.StatusBadRequest)
+				return wr, err
 			default:
 				h.logger.Error("Error while remote writing the v1 request", "err", err.Error())
```

Comment:
No need to log on 500 errors, it's done for us:

<img width="953" height="527" alt="Image" src="https://github.com/user-attachments/assets/0ae287af-6502-494e-9e18-dc89fdc920da" />